oximeter/collector/src/agent.rs

```diff
@@ -633,6 +697,123 @@ impl OximeterAgent {
+    ) -> usize {
+        let mut tasks = self.collection_tasks.lock().await;
+
+        // First prune unwanted collection tasks.
+        //
+        // This is set of all producers that we currently have, which are not in
+        // the new list from Nexus.
+        let ids_to_prune: Vec<_> = tasks
+            .keys()
+            .filter(|id| !expected_producers.contains_key(id))
+            .copied()
+            .collect();
+        let n_pruned = ids_to_prune.len();
+        for id in ids_to_prune.into_iter() {
+            // This method only returns an error if the provided ID does not
+            // exist in the current tasks. That is impossible, because we hold
+            // the lock, and we've just computed this as the set that _is_ in
+            // the map, and not in the new set from Nexus.
+            self.delete_producer_locked(&mut tasks, id).await.unwrap();
+        }
+
+        // And then ensure everything in the list.
+        //
+        // This will insert new tasks, and update any that we already know
+        // about.
+        for info in expected_producers.into_values() {
+            self.register_producer_locked(&mut tasks, info).await;
+        }
+        n_pruned
+    }
+}
+
+// A task which periodically updates our list of producers from Nexus.
+async fn refresh_producer_list(agent: OximeterAgent, resolver: Resolver) {
+    let mut interval = tokio::time::interval(agent.refresh_interval);
+    interval.tick().await; // Completes immediately.
+    let page_size = Some(NonZeroU32::new(100).unwrap());
+    loop {
+        interval.tick().await;
+        info!(agent.log, "refreshing list of producers from Nexus");
+        let nexus_addr =
+            resolve_nexus_with_backoff(&agent.log, &resolver).await;
+        let url = format!("http://{}", nexus_addr);
+        let client = nexus_client::Client::new(&url, agent.log.clone());
+        let mut stream = client.cpapi_assigned_producers_list_stream(
+            &agent.id,
+            page_size,
+            Some(IdSortMode::IdAscending),
+        );
+        let mut expected_producers = BTreeMap::new();
+        loop {
+            match stream.try_next().await {
+                Err(e) => {
+                    error!(
+                        agent.log,
+                        "error fetching next assigned producer";
+                        "err" => ?e,
+                    );
+                }
+                Ok(Some(p)) => {
+                    let endpoint = ProducerEndpoint::from(p);
+                    let old = expected_producers.insert(endpoint.id, endpoint);
+                    assert!(old.is_none());
+                }
+                Ok(None) => break,
+            }
+        }
+        let n_current_tasks = expected_producers.len();
+        let n_pruned_tasks = agent.ensure_producers(expected_producers).await;
+        let _ = agent.last_refresh_time.lock().unwrap().insert(Utc::now());
```

Comment:
Tiny style nitpick, feel free to ignore: I forget `Option` even has an `insert` method, and since we're dealing with `BTreeMap` above, I'd maybe write this as

```suggestion
        *agent.last_refresh_time.lock().unwrap() = Some(Utc::now());
```

Reply:
Done in f44da6fc0